### PR TITLE
feat: delayed write operations, ie. porcelain

### PR DESCRIPTION
**Pull request by jsam**
closes: #1870 

/deploy

### CODE REVIEW

File: renku/service/jobs/delayed_ctrl.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+from renku.service.logger import worker_log
+from renku.service.views.decorators import requires_cache
+
+
+def to_ctrl(cache, context, user_data, renku_module, renku_ctrl):
+    """Based on renku operation return valid controller."""
+    ctrl_module = importlib.import_module(renku_module)
+    cls = getattr(ctrl_module, renku_ctrl)
+
+    _ = context.pop("is_delayed", None)  # NOTE: Prevent recursive invocation.
```
Comment (Panaetius) [inline renku/service/jobs/delayed_ctrl.py:30]:
```suggestion
    context.pop("is_delayed", None)  # NOTE: Prevent recursive invocation.
```

File: renku/service/controllers/api/mixins.py [hunk 2]
```diff
@@ -128,7 +219,7 @@ def remote(self):
             return self.renku_op()
 
 
-class ReadWithSyncOperation(ReadOperationMixin, metaclass=ABCMeta):
+class RenkuOpSyncMixin(RenkuOperationMixin, metaclass=ABCMeta):
```
Comment (Panaetius) [inline renku/service/controllers/api/mixins.py:222]: Would be nice if this had a more verbose description as well

**Review by Panaetius — APPROVED** [review]

**Comment by rokroskar** [comment]
Since this is supposed to be a high-level API, is it documented somewhere that it's easy to find?

**Comment by Panaetius** [comment]
> Since this is supposed to be a high-level API, is it documented somewhere that it's easy to find?

Only in the swagger docs. It's not really a different api, just an extension of existing endpoints to improve usability.

**Comment by Panaetius** [comment]
I created https://github.com/SwissDataScienceCenter/renku-python/issues/2016 for docs

**Review by mohammad-sdsc — DISMISSED** [review]
LGTM, thank you!

File: renku/service/controllers/api/mixins.py [hunk 3]
```diff
@@ -203,27 +222,43 @@ def execute_op(self):
...
+        # NOTE: Only do a fetch every >30s to get eventual consistency but not slow things down too much
+        if project.fetch_age < PROJECT_FETCH_TIME:
+            return
+
+        lock = project.write_lock()
+
+        if self.skip_lock:
+            lock = contextlib.suppress()
+        try:
+            with lock:
```
Comment (mohammad-sdsc) [inline renku/service/controllers/api/mixins.py:235]: Should we again check if `project.fetch_age < PROJECT_FETCH_TIME` again here and do nothing if it's the case? This makes sense if multiple operations are waiting for the lock to do the reset.

**Review by mohammad-sdsc — DISMISSED** [review]
Thanks! This looks great! I've made a comment for a possible optimization; feel free to ignore it if it's not applicable.

**Review by mohammad-sdsc — APPROVED** [review]
🚀

**Pull request merged**